Cross build go-build and push manifest

mkumatag opened the pull request:
This will build all supported architectures(amd64, arm64 and ppc64le) docker images on x86 platform and push manifest for the same.

1. Use multistage build (Required Docker 17.05 or higher)
2. Included following make targets:

  - `all-push` - for pushing docker images for all the supported architecture
  - `all-build` - docker docker image for all the supported architecture
  - `push-manifest` - push manifest image, `all-push` should be run before this target.

The typical release should run - `make all-push push-manifest`

mkumatag commented:
/cc @fasaxc @djlwilder @tomdee @deitch

deitch commented on Makefile:
```diff
@@ -13,44 +13,81 @@ endif
 ifeq ($(ARCH),x86_64)
         override ARCH=amd64
 endif
+ifeq ($(ARCH),ppc64le)
```
What is the purpose of this? It just says, "if ARCH == 'ppc64le' then set ti 'ppc64le'"

mkumatag replied:
this one should be removed.

deitch commented:
Thanks for adding to the multi-platform effort, @mkumatag . 

I am confused as to how this whole process works. Before, it was pretty straightforward: one `Dockerfile` per architecture, it built the local one for your `ARCH`. 

This appears to try and do some `sed` magic on the `Dockerfile` per arch to put it in with `CROSS_BUILD_COPY qemu-ARCH-static`, which makes the `Dockerfile` on its own unusable, and creates all sorts of local temp dirs.

We also don't need qemu installed on `ppc64le` or `aarch64`, since we will be executing cross-builds only from amd64, and we already have it [here](https://github.com/projectcalico/go-build/blob/master/Dockerfile.amd64#L66). Is it so that we can build the cross-image, e.g. `calico/go-build:latest-arm64` from on an `amd64`? If so, we shouldn't actually put qemu in the final image. 

I like the idea of having a single target that when run on `amd64` creates the `go-build` images for all supported platforms - that was my next goal here (once the [felix PR](https://github.com/projectcalico/felix/pull/1703) gets merged in). This feels like it has a lot of overhead wrapping.

I am not a maintainer here, so I don't have a say, but I think this should be done in 2 steps: 

1. Cross-build and push
2. Add manifest

The manifest command has been added to the docker CLI, so using Phil's `manifest-tool` separately should be unnecessary. 

In any case, let's wait for some of the maintainers to jump in. Definitely glad we are getting others involved! :-)

djlwilder commented:
@mkumatag thanks for the great PR.
 
(Please correct me if I got this wrong) The feature added by this PR  that our previous attempt did not is to create a single "fat" calico/go-build image.  It builds a amd64, arm64 and ppc64le images all on a amd64 platform,  qemu is required for the arm and ppc images so the docker build will work.  This is exactly what we are doing for felix.

I attempted to do this earlier and found that when building arm and ppc images qemu would give an exception or hang when running go install as part of the docker build.  I  see the same problem with this PR.  When I ran "make all" it failed during the arm64 or ppc64le docker builds. Is this a problem in my environment possibly?.    Have you had better luck?

Note:  this is not a problem for the felix build as we don't run any go commands during the image builds.

mkumatag commented:
> @mkumatag thanks for the great PR.
> 
your understanding is right!
> (Please correct me if I got this wrong) The feature added by this PR that our previous attempt did not is to create a single "fat" calico/go-build image. It builds a amd64, arm64 and ppc64le images all on a amd64 platform, qemu is required for the arm and ppc images so the docker build will work. This is exactly what we are doing for felix.
> 
This worked for me without any issues, I tested in ubuntu 16.04 latest version.

```
root@jupiter-vm875:~# docker images
REPOSITORY                   TAG                 IMAGE ID            CREATED             SIZE
calico/go-build              latest-ppc64le      ce397bcb1e5d        23 minutes ago      880 MB
calico/go-build              latest-arm64        e7c6fb41c06a        30 minutes ago      1.12 GB
calico/go-build              latest-amd64        d3afb1b19470        36 minutes ago      1.04 GB
multiarch/qemu-user-static   register            114884e3e8f5        2 weeks ago         1.16 MB
arm64v8/golang               1.9.4-alpine3.7     f6e7e1435eb5        4 weeks ago         259 MB
ppc64le/golang               1.9.4-alpine3.7     23dd3507f936        4 weeks ago         262 MB
amd64/golang                 1.9.4-alpine3.7     fb6e10bf973b        4 weeks ago         269 MB
root@jupiter-vm875:~# docker run -ti calico/go-build:latest-ppc64le go version
Starting with UID : 9001
go version go1.9.4 linux/ppc64le
root@jupiter-vm875:~# docker run -ti calico/go-build:latest-amd64 go version
Starting with UID : 9001
go version go1.9.4 linux/amd64
root@jupiter-vm875:~# docker run -ti calico/go-build:latest-arm64 go version
Starting with UID : 9001
go version go1.9.4 linux/arm64
root@jupiter-vm875:~# uname -a
Linux jupiter-vm875 4.4.0-116-generic #140-Ubuntu SMP Mon Feb 12 21:23:04 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
root@jupiter-vm875:~#
```
> I attempted to do this earlier and found that when building arm and ppc images qemu would give an exception or hang when running go install as part of the docker build. I see the same problem with this PR. When I ran "make all" it failed during the arm64 or ppc64le docker builds. Is this a problem in my environment possibly?. Have you had better luck?
> 
> Note: this is not a problem for the felix build as we don't run any go commands during the image builds.

mkumatag commented:
> Thanks for adding to the multi-platform effort, @mkumatag .
> 
> I am confused as to how this whole process works. Before, it was pretty straightforward: one Dockerfile per architecture, it built the local one for your ARCH.
> 
This is mainly for copying the appropriate qemu static binary runtime into the Dockerfile to build for any architecture on amd64. Actually, most of this code is copied from kubernetes community and there we have only one Dockerfile for all the architecture and that is the reason of this whole sed part. I can simplify or merge everything into one Dockerfile if no difference in the content except BASEIMAGE.
> This appears to try and do some sed magic on the Dockerfile per arch to put it in with CROSS_BUILD_COPY qemu-ARCH-static, which makes the Dockerfile on its own unusable, and creates all sorts of local temp dirs.
> 
I'm copying qemu into other arch Dockerfile only to build those docker images on amd64, I can remove the same but will not make much difference if docker image is running in the target architecture.(e.g arm64 image running in arm64 host)
> We also don't need qemu installed on ppc64le or aarch64, since we will be executing cross-builds only from amd64, and we already have it here. Is it so that we can build the cross-image, e.g. calico/go-build:latest-arm64 from on an amd64? If so, we shouldn't actually put qemu in the final image.
> 
> I like the idea of having a single target that when run on amd64 creates the go-build images for all supported platforms - that was my next goal here (once the felix PR gets merged in). This feels like it has a lot of overhead wrapping.
> 
> I am not a maintainer here, so I don't have a say, but I think this should be done in 2 steps:
> 
> Cross-build and push
> Add manifest
docker manifest just now GA'ed and available as experimental CLI in v18.02.0-ce release, which I doubt anyone will have that docker version installed. Having said that I'm open to both the option keeping in mind that we need `v18.02.0-ce` and experimental option enabled to run manifest
> The manifest command has been added to the docker CLI, so using Phil's manifest-tool separately should be unnecessary.
> 
> In any case, let's wait for some of the maintainers to jump in. Definitely glad we are getting others involved! :-)

deitch commented:
Thanks for all the comments @mkumatag . I realized afterwards that the long comment I made might have come down like a ton of critical bricks. If so, it was unintentional.

Reading now...

deitch commented:
> This is mainly for copying the appropriate qemu static binary runtime into the Dockerfile to build for any architecture on amd64

Do you mean, copy the binary into the, e.g. `calico/go-build:latest-arm64`? The `amd64` one already has it. Since the `go-build` tagged with `arm64` is meant for building on `arm64`, you won't really need it in there, I think.

> Actually, most of this code is copied from kubernetes community and there we have only one Dockerfile for all the architecture and that is the reason of this whole sed part

The whole `sed` thing was necessary pre-multi-stage `Dockerfile` builds. If you look at the PR for creating cross-compile on `calico/felix`, you see a much easier way to do it. For example, here is the `Dockerfile.arm64`

```
FROM alpine:3.7 as qemu

ARG QEMU_VERSION=2.9.1-1
ARG CROSS_ARCHS="aarch64 ppc64le"

RUN apk --update add curl

# Enable non-native runs on amd64 architecture hosts
RUN for i in ${CROSS_ARCHS}; do curl -L https://github.com/multiarch/qemu-user-static/releases/download/v${QEMU_VERSION}/qemu-${i}-static.tar.gz | tar zxvf - -C /usr/bin; done
RUN chmod +x /usr/bin/qemu-*

FROM arm64v8/alpine:3.6 as base
MAINTAINER Shaun Crampton <shaun@tigera.io>

# Enable non-native builds of this image on an amd64 hosts.
# This must be the first RUN command in this file!
# we only need this for the intermediate "base" image, so we can run all the apk and other commands
COPY --from=qemu /usr/bin/qemu-*-static /usr/bin/
```

Everything after that will just run whatever the architecture.

Do you want a gist with your PR for the `Dockerfile` modified for that? It would make the entire install dependency encapsulated in the `Dockerfile`s.

> I'm copying qemu into other arch Dockerfile only to build those docker images on amd64

Doesn't it work already? I have done a `make calico/go-build ARCH=arm64` from an `amd64` machine and I get the correct image, assuming I have done a `make register` first (well, I don't need to since I use docker-for-mac and d4m has these archs pre-registered, but that is just d4m).

> This is the approach most of the opensource projects are taking

Some are. LinuxKit, moby, and some of the others have converted to multi-stage builds. It is cleaner and much easier to maintain.

> I'm copying qemu into other arch Dockerfile only to build those docker images on amd64, I can remove the same but will not make much difference

Actually, even using the multi-stage `Dockerfile` approach above, we still have to install `qemu-user-static` for the right arch. The above example depends upon installing it into an alpine base with each one. It would be *much* better with your approach, if we just install it `go-build`, since everything depends upon that. A slight tweak to that other PR that would make a big difference in maintainability. Nice. 

> Docker manifest just now GA'ed and available as experimental CLI in v18.02.0-ce release, which I doubt anyone will have that docker version installed

No, you are right. It is better to use the reasonably stable manifest-tool. I was involved peripherally in that one, it isn't fully there. 


Summary: 

1. nice one, sorry for having come across too critically before. I should not comment on PRs late night.
2. I will upload a gist with `Dockerfile` following that pattern, which should greatly simplify your `Makefile`.
3. I agree about using `manifest-tool` for now.
4. I still think we should keep the `manifest-tool` out of this PR to make this one just use the existing functional cross-compile and push out all the tagged images, and then do a separate PR for multi-arch manifests. 
5. I still don't understand what is missing in terms of cross-building. I can do `make calico/go-build` on any arch and it builds itself, and `make calico/go-build ARCH=arm64` or `ARCH=ppc64le` and it builds for that arch.

deitch commented:
Try this gist for `Dockerfile.arm64` https://gist.github.com/deitch/52764af38c8f3d23039f52cf9cde9111

If you use that, all of the qemu stuff in the `Makefile` goes away, all of the `sed`, all of the temporary directory stuff.

mkumatag commented:
The following gist does not have multistage building anyways!, btw: I liked the approach of having multistaged dockerfile for downloading qemu and using it. Parallelly I was thinking about having only one `Dockerfile` for all the architecture to avoid any maintenance(I don't see any difference in these dockerfiles except `FROM`). So let me see whether I can keep one `Dockerfile` for all the archs and use multistate stuff to avoid downloading qemu outside etc..

> Try this gist for Dockerfile.arm64 https://gist.github.com/deitch/52764af38c8f3d23039f52cf9cde9111
> 
> If you use that, all of the qemu stuff in the Makefile goes away, all of the sed, all of the temporary directory stuff.

deitch commented:
> The following gist does not have multistage building anyways!

Haha! Yeah, after I wrote it, I realized that only a consuming stage (like `felix` or `typha`) would need multistage, but this can be a straight `curl`.

> So let me see whether I can keep one Dockerfile for all the archs and use multistate stuff to avoid downloading qemu outside

May I suggest solving one problem at a time per PR? I believe cross-compile works. I can build `calico/go-build:latest-arm64` on my `amd64` without a problem; same for `ppc64le`. What is missing is (feel free to add):

* single target to build them all, which is as simple as building a single multiple build target
* qemu inside the `go-build` container to simplify the multistage downstream for `felix`, `typha`, etc.
* single `Dockerfile` (which I am not convinced is absolutely necessary, but who am I to argue with cleaner)
* multi-arch manifest

Why not tackle each with a simpler PR?

Sorry to keep harping on it, but is this just to hit the first 2? Or is this something not functioning with cross-compile?

mkumatag commented:
@deitch I've modified the code with following

1. Use multistage build (Required Docker 17.05 or higher)
2. Included following make targets:

  - `all-push` - for pushing docker images for all the supported architecture
  - `all-build` - docker docker image for all the supported architecture
  - `push-manifest` - push manifest image, `all-push` should be run before this target.

The typical release should run - `make all-push push-manifest`

deitch commented:
Heh, @mkumatag , I am questioning your wisdom if you are taking _me_ so seriously! :-)

In all seriousness, this is a nice piece of work. Need to get @fasaxc to sign off and merge in. First should figure out why semaphore is failing. Looks like it isn't actually building `hello-s390x`? Can you figure it out?

mkumatag commented:
@fasaxc any updates on this?

fasaxc commented:
@mkumatag Sorry, I've had a hectic couple of weeks but I haven't forgotten about this.  Looks like some conflicts have crept in, please can you rebase?

Also, semaphore seems to be failing, it runs `make test` on amd64, please can you make sure that is passing (probably be disabling the s390x test, which doesn't seem to work)?

mkumatag commented:
@fasaxc rebased now and even tests are passing.

tomdee commented:
@mkumatag This looks like great work! I tried running it locally (on amd64) but found that both the ppc64le and arm64 images builds are failing. Would you expect them to work?

<details><summary>ppc log</summary> 
```
make build ARCH=ppc64le
# Make sure we re-pull the base image to pick up security fixes.
docker build --pull -t calico/go-build:latest-ppc64le -f Dockerfile.ppc64le .
Sending build context to Docker daemon  18.78MB
Step 1/29 : FROM alpine:3.7 as qemu
3.7: Pulling from library/alpine
Digest: sha256:7df6db5aa61ae9480f52f0b3a06a140ab98d427f86d8d5de0bedab9b8df6b1c0
Status: Image is up to date for alpine:3.7
 ---> 3fd9065eaf02
Step 2/29 : ARG QEMU_VERSION=2.9.1-1
 ---> Using cache
 ---> 6ba6fa7d3532
Step 3/29 : ARG CROSS_ARCHS="aarch64 ppc64le"
 ---> Using cache
 ---> 8691e3614c5d
Step 4/29 : RUN apk --update add curl
 ---> Using cache
 ---> 2102a0bb93ba
Step 5/29 : RUN for i in ${CROSS_ARCHS}; do curl -L https://github.com/multiarch/qemu-user-static/releases/download/v${QEMU_VERSION}/qemu-${i}-static.tar.gz | tar zxvf - -C /usr/bin; done
 ---> Using cache
 ---> 0a3e332a6549
Step 6/29 : RUN chmod +x /usr/bin/qemu-*
 ---> Using cache
 ---> 7f48eb99db25
Step 7/29 : FROM ppc64le/golang:1.10.1-alpine3.7
1.10.1-alpine3.7: Pulling from ppc64le/golang
0da653ea85b5: Pull complete 
9fd90b777cc3: Pull complete 
3a20001eec33: Pull complete 
a00051ece2ae: Pull complete 
4e22c0dc3dea: Pull complete 
945f037428a6: Pull complete 
Digest: sha256:f05b355610fb5f3243d43bb03e999177b5ab4ed33774b73b85d53afd530e588d
Status: Downloaded newer image for ppc64le/golang:1.10.1-alpine3.7
 ---> 834c4e005a3a
Step 8/29 : MAINTAINER David Wilder <wilder@ibm.com>
 ---> Running in 700590f92d8e
Removing intermediate container 700590f92d8e
 ---> 3d1fcb7347eb
Step 9/29 : COPY --from=qemu /usr/bin/qemu-*-static /usr/bin/
 ---> eca73f262613
Step 10/29 : RUN apk add --no-cache su-exec curl bash git openssh mercurial make wget util-linux tini
 ---> Running in 2ed21b83d9d4
fetch http://dl-cdn.alpinelinux.org/alpine/v3.7/main/ppc64le/APKINDEX.tar.gz
fetch http://dl-cdn.alpinelinux.org/alpine/v3.7/community/ppc64le/APKINDEX.tar.gz
(1/35) Installing pkgconf (1.3.10-r0)
(2/35) Installing ncurses-terminfo-base (6.0_p20171125-r0)
(3/35) Installing ncurses-terminfo (6.0_p20171125-r0)
(4/35) Installing ncurses-libs (6.0_p20171125-r0)
(5/35) Installing readline (7.0.003-r0)
(6/35) Installing bash (4.4.19-r1)
Executing bash-4.4.19-r1.post-install
(7/35) Installing libssh2 (1.8.0-r2)
(8/35) Installing libcurl (7.59.0-r0)
(9/35) Installing curl (7.59.0-r0)
(10/35) Installing expat (2.2.5-r0)
(11/35) Installing pcre2 (10.30-r0)
(12/35) Installing git (2.15.0-r1)
(13/35) Installing make (4.2.1-r0)
(14/35) Installing libbz2 (1.0.6-r6)
(15/35) Installing libffi (3.2.1-r4)
(16/35) Installing gdbm (1.13-r1)
(17/35) Installing sqlite-libs (3.21.0-r0)
(18/35) Installing python2 (2.7.14-r2)
(19/35) Installing mercurial (4.4.1-r0)
(20/35) Installing openssh-keygen (7.5_p1-r8)
(21/35) Installing openssh-client (7.5_p1-r8)
(22/35) Installing openssh-sftp-server (7.5_p1-r8)
(23/35) Installing openssh-server-common (7.5_p1-r8)
(24/35) Installing openssh-server (7.5_p1-r8)
(25/35) Installing openssh (7.5_p1-r8)
(26/35) Installing su-exec (0.2-r0)
(27/35) Installing tini (0.16.1-r0)
(28/35) Installing libuuid (2.31-r0)
(29/35) Installing libblkid (2.31-r0)
(30/35) Installing libmount (2.31-r0)
(31/35) Installing libsmartcols (2.31-r0)
(32/35) Installing findmnt (2.31-r0)
(33/35) Installing libfdisk (2.31-r0)
(34/35) Installing util-linux (2.31-r0)
(35/35) Installing wget (1.19.2-r1)
Executing busybox-1.27.2-r6.trigger
OK: 91 MiB in 47 packages
Removing intermediate container 2ed21b83d9d4
 ---> 6b73b7424080
Step 11/29 : RUN apk upgrade --no-cache
 ---> Running in 6a9d0eae1649
fetch http://dl-cdn.alpinelinux.org/alpine/v3.7/main/ppc64le/APKINDEX.tar.gz
fetch http://dl-cdn.alpinelinux.org/alpine/v3.7/community/ppc64le/APKINDEX.tar.gz
Upgrading critical system libraries and apk-tools:
(1/1) Upgrading apk-tools (2.8.1-r1 -> 2.9.1-r2)
Executing busybox-1.27.2-r6.trigger
Continuing the upgrade transaction with new apk-tools:
fetch http://dl-cdn.alpinelinux.org/alpine/v3.7/main/ppc64le/APKINDEX.tar.gz
fetch http://dl-cdn.alpinelinux.org/alpine/v3.7/community/ppc64le/APKINDEX.tar.gz
(1/1) Upgrading busybox (1.27.2-r6 -> 1.27.2-r8)
Executing busybox-1.27.2-r8.post-upgrade
Executing busybox-1.27.2-r8.trigger
OK: 91 MiB in 47 packages
Removing intermediate container 6a9d0eae1649
 ---> 7d87996848e3
Step 12/29 : RUN echo 'Host *' >> /etc/ssh/ssh_config
 ---> Running in 92a29dea6e5a
Removing intermediate container 92a29dea6e5a
 ---> b1c5e3ff6375
Step 13/29 : RUN echo '    StrictHostKeyChecking no' >> /etc/ssh/ssh_config
 ---> Running in 42d8bc6c1fe2
Removing intermediate container 42d8bc6c1fe2
 ---> acc7102688e8
Step 14/29 : ENV CGO_ENABLED=0
 ---> Running in cfb5e71aa238
Removing intermediate container cfb5e71aa238
 ---> 3d2468d458f6
Step 15/29 : RUN go install -v std
 ---> Running in 9fc8d90ef1a2
plugin
os/user
net
internal/testenv
archive/tar
vendor/golang_org/x/net/lex/httplex
vendor/golang_org/x/net/proxy
vendor/golang_org/x/net/internal/nettest
log/syslog
vendor/golang_org/x/net/nettest
net/textproto
crypto/x509
crypto/tls
net/http/httptrace
net/smtp
net/http
net/http/cgi
net/http/cookiejar
net/http/httptest
net/http/httputil
net/rpc
net/http/fcgi
net/rpc/jsonrpc
Removing intermediate container 9fc8d90ef1a2
 ---> 474d25251188
Step 16/29 : RUN go get github.com/Masterminds/glide
 ---> Running in 2017cc4784b3
fatal error: unexpected signal during runtime execution
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x100351f4]

runtime stack:
syscall.Syscall6(0xc420216060, 0x10005898, 0xc5, 0xc420220020, 0x20, 0x104547e0, 0xc420216060, 0x0, 0xc5, 0x0)
	/usr/local/go/src/syscall/asm_linux_ppc64x.s:61 +0x6c
os.(*Process).blockUntilWaitable(0x3, 0xc420220020, 0x0, 0xc420216060)
	/usr/local/go/src/os/wait_waitid.go:31 +0x8c
runtime: unexpected return pc for runtime.(*mcache).nextFree.func1 called from 0xc4202ba058
stack: frame={sp:0xc420350970, fp:0xc420350970} stack=[0x40007edbe0,0x4000800c50)


goroutine 1 [semacquire]:
sync.runtime_Semacquire(0xc42022f99c)
	/usr/local/go/src/runtime/sema.go:56 +0x38
sync.(*WaitGroup).Wait(0xc42022f990)
	/usr/local/go/src/sync/waitgroup.go:129 +0x9c
cmd/go/internal/work.(*Builder).Do(0xc4201f6640, 0xc42051fa40)
	/usr/local/go/src/cmd/go/internal/work/exec.go:173 +0x3b4
cmd/go/internal/work.InstallPackages(0xc420230850, 0x1, 0x1, 0xc4200bc401)
	/usr/local/go/src/cmd/go/internal/work/build.go:482 +0xafc
cmd/go/internal/get.runGet(0x106a9480, 0xc420230850, 0x1, 0x1)
	/usr/local/go/src/cmd/go/internal/get/get.go:164 +0x1f4
main.main()
	/usr/local/go/src/cmd/go/main.go:140 +0x6f0

goroutine 19 [syscall]:
os/signal.signal_recv(0x0)
	/usr/local/go/src/runtime/sigqueue.go:139 +0xfc
os/signal.loop()
	/usr/local/go/src/os/signal/signal_unix.go:22 +0x24
created by os/signal.init.0
	/usr/local/go/src/os/signal/signal_unix.go:28 +0x3c

goroutine 66 [runnable]:
fatal error: unexpected signal during runtime execution
panic during panic
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1004ece0]

runtime stack:
syscall.Syscall6(0xc420216060, 0x10005898, 0xc5, 0xc420220020, 0x20, 0x104547e0, 0xc420216060, 0x0, 0xc5, 0x0)
	/usr/local/go/src/syscall/asm_linux_ppc64x.s:61 +0x6c
os.(*Process).blockUntilWaitable(0x3, 0xc420220020, 0x0, 0xc420216060)
	/usr/local/go/src/os/wait_waitid.go:31 +0x8c
runtime: unexpected return pc for runtime.(*mcache).nextFree.func1 called from 0xc4202ba058
stack: frame={sp:0xc420350970, fp:0xc420350970} stack=[0x40007edbe0,0x4000800c50)

The command '/bin/sh -c go get github.com/Masterminds/glide' returned a non-zero code: 2
Makefile:54: recipe for target 'calico/go-build' failed
make: *** [calico/go-build] Error 2
```
</details>
<details><summary>aarm64 log</summary> 
```
make build ARCH=arm64
make[1]: Entering directory '/home/tom/go/src/github.com/projectcalico/go-build'
# Make sure we re-pull the base image to pick up security fixes.
docker build --pull -t calico/go-build:latest-arm64 -f Dockerfile.arm64 .
Sending build context to Docker daemon  18.78MB
Step 1/28 : FROM alpine:3.7 as qemu
3.7: Pulling from library/alpine
Digest: sha256:7df6db5aa61ae9480f52f0b3a06a140ab98d427f86d8d5de0bedab9b8df6b1c0
Status: Image is up to date for alpine:3.7
 ---> 3fd9065eaf02
Step 2/28 : ARG QEMU_VERSION=2.9.1-1
 ---> Running in 7ff1796dc678
Removing intermediate container 7ff1796dc678
 ---> 6ba6fa7d3532
Step 3/28 : ARG CROSS_ARCHS="aarch64 ppc64le"
 ---> Running in a589bdaed3e0
Removing intermediate container a589bdaed3e0
 ---> 8691e3614c5d
Step 4/28 : RUN apk --update add curl
 ---> Running in 30aab3fceec9
fetch http://dl-cdn.alpinelinux.org/alpine/v3.7/main/x86_64/APKINDEX.tar.gz
fetch http://dl-cdn.alpinelinux.org/alpine/v3.7/community/x86_64/APKINDEX.tar.gz
(1/4) Installing ca-certificates (20171114-r0)
(2/4) Installing libssh2 (1.8.0-r2)
(3/4) Installing libcurl (7.59.0-r0)
(4/4) Installing curl (7.59.0-r0)
Executing busybox-1.27.2-r7.trigger
Executing ca-certificates-20171114-r0.trigger
OK: 6 MiB in 15 packages
Removing intermediate container 30aab3fceec9
 ---> 2102a0bb93ba
Step 5/28 : RUN for i in ${CROSS_ARCHS}; do curl -L https://github.com/multiarch/qemu-user-static/releases/download/v${QEMU_VERSION}/qemu-${i}-static.tar.gz | tar zxvf - -C /usr/bin; done
 ---> Running in 22f749c8dabe
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   616    0   616    0     0   2248      0 --:--:-- --:--:-- --:--:--  2256
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0qemu-aarch64-static
100 1346k  100 1346k    0     0   778k      0  0:00:01  0:00:01 --:--:-- 1078k
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   616    0   616    0     0   1844      0 --:--:-- --:--:-- --:--:--  1844
  0 1439k    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0qemu-ppc64le-static
100 1439k  100 1439k    0     0   790k      0  0:00:01  0:00:01 --:--:-- 1319k
Removing intermediate container 22f749c8dabe
 ---> 0a3e332a6549
Step 6/28 : RUN chmod +x /usr/bin/qemu-*
 ---> Running in 3a5b7f3997f5
Removing intermediate container 3a5b7f3997f5
 ---> 7f48eb99db25
Step 7/28 : FROM arm64v8/golang:1.10.1-alpine3.7
1.10.1-alpine3.7: Pulling from arm64v8/golang
b78042c299ad: Pull complete 
6fd45b97b6c2: Pull complete 
660093e5ae1b: Pull complete 
e8a7908d7c13: Pull complete 
5b5522948e3b: Pull complete 
59ce1d4b6432: Pull complete 
Digest: sha256:ff2296751188bed052d56b6e5ae4c8fa86f812c7e29cdb14dbab9fc2bbd18fd4
Status: Downloaded newer image for arm64v8/golang:1.10.1-alpine3.7
 ---> 826c89a5e187
Step 8/28 : MAINTAINER Trevor Tao <trevor.tao@arm.com>
 ---> Running in 1e5c9702e62c
Removing intermediate container 1e5c9702e62c
 ---> d6f95ccaf222
Step 9/28 : COPY --from=qemu /usr/bin/qemu-*-static /usr/bin/
 ---> 75e8457275b2
Step 10/28 : RUN apk add --no-cache su-exec curl bash git openssh mercurial make wget util-linux docker tini
 ---> Running in 5c721b07feea
fetch http://dl-cdn.alpinelinux.org/alpine/v3.7/main/aarch64/APKINDEX.tar.gz
fetch http://dl-cdn.alpinelinux.org/alpine/v3.7/community/aarch64/APKINDEX.tar.gz
(1/43) Installing pkgconf (1.3.10-r0)
(2/43) Installing ncurses-terminfo-base (6.0_p20171125-r0)
(3/43) Installing ncurses-terminfo (6.0_p20171125-r0)
(4/43) Installing ncurses-libs (6.0_p20171125-r0)
(5/43) Installing readline (7.0.003-r0)
(6/43) Installing bash (4.4.19-r1)
Executing bash-4.4.19-r1.post-install
(7/43) Installing libssh2 (1.8.0-r2)
(8/43) Installing libcurl (7.59.0-r0)
(9/43) Installing curl (7.59.0-r0)
(10/43) Installing libmnl (1.0.4-r0)
(11/43) Installing jansson (2.10-r0)
(12/43) Installing libnftnl-libs (1.0.8-r1)
(13/43) Installing iptables (1.6.1-r1)
(14/43) Installing device-mapper-libs (2.02.175-r0)
(15/43) Installing libltdl (2.4.6-r4)
(16/43) Installing libseccomp (2.3.2-r1)
(17/43) Installing docker (17.12.1-r0)
Executing docker-17.12.1-r0.pre-install
(18/43) Installing expat (2.2.5-r0)
(19/43) Installing pcre2 (10.30-r0)
(20/43) Installing git (2.15.0-r1)
(21/43) Installing make (4.2.1-r0)
(22/43) Installing libbz2 (1.0.6-r6)
(23/43) Installing libffi (3.2.1-r4)
(24/43) Installing gdbm (1.13-r1)
(25/43) Installing sqlite-libs (3.21.0-r0)
(26/43) Installing python2 (2.7.14-r2)
(27/43) Installing mercurial (4.4.1-r0)
(28/43) Installing openssh-keygen (7.5_p1-r8)
(29/43) Installing openssh-client (7.5_p1-r8)
(30/43) Installing openssh-sftp-server (7.5_p1-r8)
(31/43) Installing openssh-server-common (7.5_p1-r8)
(32/43) Installing openssh-server (7.5_p1-r8)
(33/43) Installing openssh (7.5_p1-r8)
(34/43) Installing su-exec (0.2-r0)
(35/43) Installing tini (0.16.1-r0)
(36/43) Installing libuuid (2.31-r0)
(37/43) Installing libblkid (2.31-r0)
(38/43) Installing libmount (2.31-r0)
(39/43) Installing libsmartcols (2.31-r0)
(40/43) Installing findmnt (2.31-r0)
(41/43) Installing libfdisk (2.31-r0)
(42/43) Installing util-linux (2.31-r0)
(43/43) Installing wget (1.19.2-r1)
Executing busybox-1.27.2-r6.trigger
OK: 196 MiB in 55 packages
Removing intermediate container 5c721b07feea
 ---> a19afa8309b6
Step 11/28 : RUN apk upgrade --no-cache
 ---> Running in 7c398d1891e0
fetch http://dl-cdn.alpinelinux.org/alpine/v3.7/main/aarch64/APKINDEX.tar.gz
fetch http://dl-cdn.alpinelinux.org/alpine/v3.7/community/aarch64/APKINDEX.tar.gz
Upgrading critical system libraries and apk-tools:
(1/1) Upgrading apk-tools (2.8.1-r1 -> 2.9.1-r0)
Executing busybox-1.27.2-r6.trigger
Continuing the upgrade transaction with new apk-tools:
fetch http://dl-cdn.alpinelinux.org/alpine/v3.7/main/aarch64/APKINDEX.tar.gz
fetch http://dl-cdn.alpinelinux.org/alpine/v3.7/community/aarch64/APKINDEX.tar.gz
(1/1) Upgrading busybox (1.27.2-r6 -> 1.27.2-r8)
Executing busybox-1.27.2-r8.post-upgrade
Executing busybox-1.27.2-r8.trigger
OK: 196 MiB in 55 packages
Removing intermediate container 7c398d1891e0
 ---> 28303ce35cb3
Step 12/28 : RUN echo 'Host *' >> /etc/ssh/ssh_config   && echo '    StrictHostKeyChecking no' >> /etc/ssh/ssh_config
 ---> Running in 2784dbeba042
Removing intermediate container 2784dbeba042
 ---> 29d1bc3288af
Step 13/28 : ENV CGO_ENABLED=0
 ---> Running in 84f22d6e5684
Removing intermediate container 84f22d6e5684
 ---> 4eed9cb6b52e
Step 14/28 : RUN go install -v std
 ---> Running in 7c4af98e9bad
plugin
os/user
net
internal/testenv
archive/tar
vendor/golang_org/x/net/lex/httplex
vendor/golang_org/x/net/proxy
log/syslog
vendor/golang_org/x/net/nettest
net/textproto
vendor/golang_org/x/net/internal/nettest
crypto/x509
crypto/tls
net/http/httptrace
net/smtp
net/http
net/http/cgi
net/http/cookiejar
net/http/httptest
net/http/httputil
net/rpc
net/http/fcgi
net/rpc/jsonrpc
# net/http/fcgi
qemu:handle_cpu_signal received signal outside vCPU context @ pc=0x7f947c02f430
qemu:handle_cpu_signal received signal outside vCPU context @ pc=0x6015ee41
The command '/bin/sh -c go install -v std' returned a non-zero code: 2
Makefile:54: recipe for target 'calico/go-build' failed
make[1]: *** [calico/go-build] Error 2
make[1]: Leaving directory '/home/tom/go/src/github.com/projectcalico/go-build'
Makefile:49: recipe for target 'sub-build-arm64' failed
make: *** [sub-build-arm64] Error 2
```
</details>

mkumatag commented:
I'm not facing any issues on my machine, I was able to build for all three arches(amd64, arm64 and ppc64le) on amd64(host arch)

```shell
root@localhost:~/go-build# docker images | grep go-build
calico/go-build              latest-ppc64le      186032facb14        9 minutes ago       1.11GB
calico/go-build              latest-arm64        2720b9ec090a        19 minutes ago      1.34GB
calico/go-build              latest-amd64        f73b84a1cff4        25 minutes ago      1.25GB
root@localhost:~/go-build# docker run -ti calico/go-build:latest-ppc64le go version
Starting with UID : 9001
go version go1.10.1 linux/ppc64le
root@localhost:~/go-build# docker run -ti calico/go-build:latest-arm64 go version
Starting with UID : 9001
go version go1.10.1 linux/arm64
root@localhost:~/go-build# docker run -ti calico/go-build:latest-amd64 go version
Starting with UID : 9001
go version go1.10.1 linux/amd64
root@localhost:~/go-build#
```

<details>
<summary>arm64 build</summary>
<br>
make build ARCH=arm64
make[1]: Entering directory '/root/go-build'
# Make sure we re-pull the base image to pick up security fixes.
docker build --pull -t calico/go-build:latest-arm64 -f Dockerfile.arm64 .
Sending build context to Docker daemon  253.4kB
Step 1/28 : FROM alpine:3.7 as qemu
3.7: Pulling from library/alpine
ff3a5c916c92: Already exists
Digest: sha256:7df6db5aa61ae9480f52f0b3a06a140ab98d427f86d8d5de0bedab9b8df6b1c0
Status: Downloaded newer image for alpine:3.7
 ---> 3fd9065eaf02
Step 2/28 : ARG QEMU_VERSION=2.9.1-1
 ---> Running in 12fc7adb8361
Removing intermediate container 12fc7adb8361
 ---> d4c99099ef6b
Step 3/28 : ARG CROSS_ARCHS="aarch64 ppc64le"
 ---> Running in f23a86c692ba
Removing intermediate container f23a86c692ba
 ---> 2b216dfd17d3
Step 4/28 : RUN apk --update add curl
 ---> Running in 9a04f39058a2
fetch http://dl-cdn.alpinelinux.org/alpine/v3.7/main/x86_64/APKINDEX.tar.gz
fetch http://dl-cdn.alpinelinux.org/alpine/v3.7/community/x86_64/APKINDEX.tar.gz
(1/4) Installing ca-certificates (20171114-r0)
(2/4) Installing libssh2 (1.8.0-r2)
(3/4) Installing libcurl (7.59.0-r0)
(4/4) Installing curl (7.59.0-r0)
Executing busybox-1.27.2-r7.trigger
Executing ca-certificates-20171114-r0.trigger
OK: 6 MiB in 15 packages
Removing intermediate container 9a04f39058a2
 ---> f363c6503369
Step 5/28 : RUN for i in ${CROSS_ARCHS}; do curl -L https://github.com/multiarch/qemu-user-static/releases/download/v${QEMU_VERSION}/qemu-${i}-static.tar.gz | tar zxvf - -C /usr/bin; done
 ---> Running in cd1a6f3bdfe9
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   616    0   616    0     0   5176      0 --:--:-- --:--:-- --:--:--  5176
qemu-aarch64-static
100 1346k  100 1346k    0     0  3461k      0 --:--:-- --:--:-- --:--:-- 3461k
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   616    0   616    0     0   5549      0 --:--:-- --:--:-- --:--:--  5549
  2 1439k    2 33323    0     0   132k      0  0:00:10 --:--:--  0:00:10  132kqemu-ppc64le-static
100 1439k  100 1439k    0     0  4113k      0 --:--:-- --:--:-- --:--:-- 13.0M
Removing intermediate container cd1a6f3bdfe9
 ---> 8605df2fbae1
Step 6/28 : RUN chmod +x /usr/bin/qemu-*
 ---> Running in d93f30c98e68
Removing intermediate container d93f30c98e68
 ---> a4ef882bf84f
Step 7/28 : FROM arm64v8/golang:1.10.1-alpine3.7
1.10.1-alpine3.7: Pulling from arm64v8/golang
b78042c299ad: Already exists
6fd45b97b6c2: Already exists
660093e5ae1b: Already exists
e8a7908d7c13: Already exists
5b5522948e3b: Already exists
59ce1d4b6432: Already exists
Digest: sha256:ff2296751188bed052d56b6e5ae4c8fa86f812c7e29cdb14dbab9fc2bbd18fd4
Status: Downloaded newer image for arm64v8/golang:1.10.1-alpine3.7
 ---> 826c89a5e187
Step 8/28 : MAINTAINER Trevor Tao <trevor.tao@arm.com>
 ---> Running in 8d572e387548
Removing intermediate container 8d572e387548
 ---> 35459753628c
Step 9/28 : COPY --from=qemu /usr/bin/qemu-*-static /usr/bin/
 ---> 4d2e6a000c14
Step 10/28 : RUN apk add --no-cache su-exec curl bash git openssh mercurial make wget util-linux docker tini
 ---> Running in 24e55bc0480f
fetch http://dl-cdn.alpinelinux.org/alpine/v3.7/main/aarch64/APKINDEX.tar.gz
fetch http://dl-cdn.alpinelinux.org/alpine/v3.7/community/aarch64/APKINDEX.tar.gz
(1/43) Installing pkgconf (1.3.10-r0)
(2/43) Installing ncurses-terminfo-base (6.0_p20171125-r0)
(3/43) Installing ncurses-terminfo (6.0_p20171125-r0)
(4/43) Installing ncurses-libs (6.0_p20171125-r0)
(5/43) Installing readline (7.0.003-r0)
(6/43) Installing bash (4.4.19-r1)
Executing bash-4.4.19-r1.post-install
(7/43) Installing libssh2 (1.8.0-r2)
(8/43) Installing libcurl (7.59.0-r0)
(9/43) Installing curl (7.59.0-r0)
(10/43) Installing libmnl (1.0.4-r0)
(11/43) Installing jansson (2.10-r0)
(12/43) Installing libnftnl-libs (1.0.8-r1)
(13/43) Installing iptables (1.6.1-r1)
(14/43) Installing device-mapper-libs (2.02.175-r0)
(15/43) Installing libltdl (2.4.6-r4)
(16/43) Installing libseccomp (2.3.2-r1)
(17/43) Installing docker (17.12.1-r0)
Executing docker-17.12.1-r0.pre-install
(18/43) Installing expat (2.2.5-r0)
(19/43) Installing pcre2 (10.30-r0)
(20/43) Installing git (2.15.0-r1)
(21/43) Installing make (4.2.1-r0)
(22/43) Installing libbz2 (1.0.6-r6)
(23/43) Installing libffi (3.2.1-r4)
(24/43) Installing gdbm (1.13-r1)
(25/43) Installing sqlite-libs (3.21.0-r0)
(26/43) Installing python2 (2.7.14-r2)
(27/43) Installing mercurial (4.4.1-r0)
(28/43) Installing openssh-keygen (7.5_p1-r8)
(29/43) Installing openssh-client (7.5_p1-r8)
(30/43) Installing openssh-sftp-server (7.5_p1-r8)
(31/43) Installing openssh-server-common (7.5_p1-r8)
(32/43) Installing openssh-server (7.5_p1-r8)
(33/43) Installing openssh (7.5_p1-r8)
(34/43) Installing su-exec (0.2-r0)
(35/43) Installing tini (0.16.1-r0)
(36/43) Installing libuuid (2.31-r0)
(37/43) Installing libblkid (2.31-r0)
(38/43) Installing libmount (2.31-r0)
(39/43) Installing libsmartcols (2.31-r0)
(40/43) Installing findmnt (2.31-r0)
(41/43) Installing libfdisk (2.31-r0)
(42/43) Installing util-linux (2.31-r0)
(43/43) Installing wget (1.19.2-r1)
Executing busybox-1.27.2-r6.trigger
OK: 196 MiB in 55 packages
Removing intermediate container 24e55bc0480f
 ---> 19ec64f5a6b4
Step 11/28 : RUN apk upgrade --no-cache
 ---> Running in 276036766fa9
fetch http://dl-cdn.alpinelinux.org/alpine/v3.7/main/aarch64/APKINDEX.tar.gz
fetch http://dl-cdn.alpinelinux.org/alpine/v3.7/community/aarch64/APKINDEX.tar.gz
Upgrading critical system libraries and apk-tools:
(1/1) Upgrading apk-tools (2.8.1-r1 -> 2.9.1-r0)
Executing busybox-1.27.2-r6.trigger
Continuing the upgrade transaction with new apk-tools:
fetch http://dl-cdn.alpinelinux.org/alpine/v3.7/main/aarch64/APKINDEX.tar.gz
fetch http://dl-cdn.alpinelinux.org/alpine/v3.7/community/aarch64/APKINDEX.tar.gz
(1/1) Upgrading busybox (1.27.2-r6 -> 1.27.2-r8)
Executing busybox-1.27.2-r8.post-upgrade
Executing busybox-1.27.2-r8.trigger
OK: 196 MiB in 55 packages
Removing intermediate container 276036766fa9
 ---> 70ad11ddc7df
Step 12/28 : RUN echo 'Host *' >> /etc/ssh/ssh_config   && echo '    StrictHostKeyChecking no' >> /etc/ssh/ssh_config
 ---> Running in b4135a275fc4
Removing intermediate container b4135a275fc4
 ---> e8ba78cebae4
Step 13/28 : ENV CGO_ENABLED=0
 ---> Running in 9d8a03d02b1b
Removing intermediate container 9d8a03d02b1b
 ---> 40d7076500c6
Step 14/28 : RUN go install -v std
 ---> Running in c37d24d548b5
os/user
archive/tar
net
internal/testenv
plugin
vendor/golang_org/x/net/lex/httplex
crypto/x509
vendor/golang_org/x/net/proxy
net/textproto
log/syslog
crypto/tls
vendor/golang_org/x/net/internal/nettest
vendor/golang_org/x/net/nettest
net/smtp
net/http/httptrace
net/http
net/http/cgi
net/http/cookiejar
net/http/httptest
net/http/fcgi
net/http/httputil
net/rpc
net/rpc/jsonrpc
Removing intermediate container c37d24d548b5
 ---> 21da6b102b15
Step 15/28 : RUN go get github.com/Masterminds/glide
 ---> Running in f8c90544e1c9
Removing intermediate container f8c90544e1c9
 ---> 218c170b9e51
Step 16/28 : ENV GLIDE_HOME /home/user/.glide
 ---> Running in 7970362dc9c1
Removing intermediate container 7970362dc9c1
 ---> 794543eb4f0d
Step 17/28 : RUN go get github.com/golang/dep/cmd/dep
 ---> Running in 351a2517f987
Removing intermediate container 351a2517f987
 ---> d20d92de1ef9
Step 18/28 : RUN go get github.com/onsi/ginkgo/ginkgo
 ---> Running in a6c2514ce5c2
Removing intermediate container a6c2514ce5c2
 ---> c7a1ed591128
Step 19/28 : RUN go get -u -d github.com/alecthomas/gometalinter &&     cd /go/src/github.com/alecthomas/gometalinter &&     git checkout cc4415ed09f7073d595ee504cad4d98b71a3038e &&     go install github.com/alecthomas/gometalinter
 ---> Running in 54ab0334b395
Note: checking out 'cc4415ed09f7073d595ee504cad4d98b71a3038e'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by performing another checkout.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -b with the checkout command again. Example:

  git checkout -b <new-branch-name>

HEAD is now at cc4415e... Add back support for predefined patterns in linter definitions.
Removing intermediate container 54ab0334b395
 ---> e9173c49e6fa
Step 20/28 : RUN ln -s `which gometalinter` /usr/local/bin/gometalinter
 ---> Running in 25b0722ea9b0
Removing intermediate container 25b0722ea9b0
 ---> a0776d76e10a
Step 21/28 : RUN gometalinter --install
 ---> Running in a1df3a0f4d29
Installing:
  deadcode
  dupl
  errcheck
  gas
  goconst
  gocyclo
  goimports
  golint
  gosimple
  gotype
  gotypex
  ineffassign
  interfacer
  lll
  maligned
  megacheck
  misspell
  safesql
  staticcheck
  structcheck
  unconvert
  unparam
  unused
  varcheck
  vet
Removing intermediate container a1df3a0f4d29
 ---> 13b200042fc6
Step 22/28 : RUN go get github.com/pmezard/licenses
 ---> Running in b0038668c991
Removing intermediate container b0038668c991
 ---> 498d7b63ebe1
Step 23/28 : RUN go get github.com/wadey/gocovmerge
 ---> Running in 5b86a0b0ee36
Removing intermediate container 5b86a0b0ee36
 ---> c35f98a8ffa7
Step 24/28 : RUN go get github.com/mikefarah/yaml
 ---> Running in 3221aeec6241
Removing intermediate container 3221aeec6241
 ---> 544dc8e836c9
Step 25/28 : RUN go get -u -d github.com/fasaxc/goveralls &&     go get golang.org/x/tools/cover &&     go get golang.org/x/net/html &&     go get golang.org/x/net/websocket &&     go get github.com/stretchr/testify/assert &&     go get golang.org/x/text/encoding &&     go get golang.org/x/crypto/ssh/terminal &&     cd /go/src/github.com/fasaxc/goveralls &&     git checkout tags/v0.0.1-smc &&     go install github.com/fasaxc/goveralls
 ---> Running in a6da4bea21cf
Note: checking out 'tags/v0.0.1-smc'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by performing another checkout.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -b with the checkout command again. Example:

  git checkout -b <new-branch-name>

HEAD is now at e305190... Only send a service job ID if it is provided.
Removing intermediate container a6da4bea21cf
 ---> 02d7dab62db6
Step 26/28 : RUN chmod -R 777 $GOPATH
 ---> Running in b8224cd1dd4f
Removing intermediate container b8224cd1dd4f
 ---> ac45c019480f
Step 27/28 : COPY entrypoint.sh /usr/local/bin/entrypoint.sh
 ---> d635bb562f3b
Step 28/28 : ENTRYPOINT ["/sbin/tini", "--", "/usr/local/bin/entrypoint.sh"]
 ---> Running in 9998287b4d6e
Removing intermediate container 9998287b4d6e
 ---> 2720b9ec090a
Successfully built 2720b9ec090a
Successfully tagged calico/go-build:latest-arm64
</details>

<details>
<summary>ppc64le build</summary>
<br>
make build ARCH=ppc64le
make[1]: Entering directory '/root/go-build'
# Make sure we re-pull the base image to pick up security fixes.
docker build --pull -t calico/go-build:latest-ppc64le -f Dockerfile.ppc64le .
Sending build context to Docker daemon  253.4kB
Step 1/29 : FROM alpine:3.7 as qemu
3.7: Pulling from library/alpine
Digest: sha256:7df6db5aa61ae9480f52f0b3a06a140ab98d427f86d8d5de0bedab9b8df6b1c0
Status: Image is up to date for alpine:3.7
 ---> 3fd9065eaf02
Step 2/29 : ARG QEMU_VERSION=2.9.1-1
 ---> Using cache
 ---> d4c99099ef6b
Step 3/29 : ARG CROSS_ARCHS="aarch64 ppc64le"
 ---> Using cache
 ---> 2b216dfd17d3
Step 4/29 : RUN apk --update add curl
 ---> Using cache
 ---> f363c6503369
Step 5/29 : RUN for i in ${CROSS_ARCHS}; do curl -L https://github.com/multiarch/qemu-user-static/releases/download/v${QEMU_VERSION}/qemu-${i}-static.tar.gz | tar zxvf - -C /usr/bin; done
 ---> Using cache
 ---> 8605df2fbae1
Step 6/29 : RUN chmod +x /usr/bin/qemu-*
 ---> Using cache
 ---> a4ef882bf84f
Step 7/29 : FROM ppc64le/golang:1.10.1-alpine3.7
1.10.1-alpine3.7: Pulling from ppc64le/golang
0da653ea85b5: Pull complete
9fd90b777cc3: Pull complete
3a20001eec33: Pull complete
a00051ece2ae: Pull complete
4e22c0dc3dea: Pull complete
945f037428a6: Pull complete
Digest: sha256:f05b355610fb5f3243d43bb03e999177b5ab4ed33774b73b85d53afd530e588d
Status: Downloaded newer image for ppc64le/golang:1.10.1-alpine3.7
 ---> 834c4e005a3a
Step 8/29 : MAINTAINER David Wilder <wilder@ibm.com>
 ---> Running in 0cf158890894
Removing intermediate container 0cf158890894
 ---> 3a974e052eca
Step 9/29 : COPY --from=qemu /usr/bin/qemu-*-static /usr/bin/
 ---> 3d1d3e993f8d
Step 10/29 : RUN apk add --no-cache su-exec curl bash git openssh mercurial make wget util-linux tini
 ---> Running in 758e4a4cce90
fetch http://dl-cdn.alpinelinux.org/alpine/v3.7/main/ppc64le/APKINDEX.tar.gz
fetch http://dl-cdn.alpinelinux.org/alpine/v3.7/community/ppc64le/APKINDEX.tar.gz
(1/35) Installing pkgconf (1.3.10-r0)
(2/35) Installing ncurses-terminfo-base (6.0_p20171125-r0)
(3/35) Installing ncurses-terminfo (6.0_p20171125-r0)
(4/35) Installing ncurses-libs (6.0_p20171125-r0)
(5/35) Installing readline (7.0.003-r0)
(6/35) Installing bash (4.4.19-r1)
Executing bash-4.4.19-r1.post-install
(7/35) Installing libssh2 (1.8.0-r2)
(8/35) Installing libcurl (7.59.0-r0)
(9/35) Installing curl (7.59.0-r0)
(10/35) Installing expat (2.2.5-r0)
(11/35) Installing pcre2 (10.30-r0)
(12/35) Installing git (2.15.0-r1)
(13/35) Installing make (4.2.1-r0)
(14/35) Installing libbz2 (1.0.6-r6)
(15/35) Installing libffi (3.2.1-r4)
(16/35) Installing gdbm (1.13-r1)
(17/35) Installing sqlite-libs (3.21.0-r0)
(18/35) Installing python2 (2.7.14-r2)
(19/35) Installing mercurial (4.4.1-r0)
(20/35) Installing openssh-keygen (7.5_p1-r8)
(21/35) Installing openssh-client (7.5_p1-r8)
(22/35) Installing openssh-sftp-server (7.5_p1-r8)
(23/35) Installing openssh-server-common (7.5_p1-r8)
(24/35) Installing openssh-server (7.5_p1-r8)
(25/35) Installing openssh (7.5_p1-r8)
(26/35) Installing su-exec (0.2-r0)
(27/35) Installing tini (0.16.1-r0)
(28/35) Installing libuuid (2.31-r0)
(29/35) Installing libblkid (2.31-r0)
(30/35) Installing libmount (2.31-r0)
(31/35) Installing libsmartcols (2.31-r0)
(32/35) Installing findmnt (2.31-r0)
(33/35) Installing libfdisk (2.31-r0)
(34/35) Installing util-linux (2.31-r0)
(35/35) Installing wget (1.19.2-r1)
Executing busybox-1.27.2-r6.trigger
OK: 91 MiB in 47 packages
Removing intermediate container 758e4a4cce90
 ---> 5c5c270a5ba5
Step 11/29 : RUN apk upgrade --no-cache
 ---> Running in 5d651655f539
fetch http://dl-cdn.alpinelinux.org/alpine/v3.7/main/ppc64le/APKINDEX.tar.gz
fetch http://dl-cdn.alpinelinux.org/alpine/v3.7/community/ppc64le/APKINDEX.tar.gz
Upgrading critical system libraries and apk-tools:
(1/1) Upgrading apk-tools (2.8.1-r1 -> 2.9.1-r2)
Executing busybox-1.27.2-r6.trigger
Continuing the upgrade transaction with new apk-tools:
fetch http://dl-cdn.alpinelinux.org/alpine/v3.7/main/ppc64le/APKINDEX.tar.gz
fetch http://dl-cdn.alpinelinux.org/alpine/v3.7/community/ppc64le/APKINDEX.tar.gz
(1/1) Upgrading busybox (1.27.2-r6 -> 1.27.2-r8)
Executing busybox-1.27.2-r8.post-upgrade
Executing busybox-1.27.2-r8.trigger
OK: 91 MiB in 47 packages
Removing intermediate container 5d651655f539
 ---> 62c033ecb0e5
Step 12/29 : RUN echo 'Host *' >> /etc/ssh/ssh_config
 ---> Running in de19f44bd5cb
Removing intermediate container de19f44bd5cb
 ---> 3fabbea05441
Step 13/29 : RUN echo '    StrictHostKeyChecking no' >> /etc/ssh/ssh_config
 ---> Running in 7216413330e6
Removing intermediate container 7216413330e6
 ---> ab8972fbd247
Step 14/29 : ENV CGO_ENABLED=0
 ---> Running in e5168f00e8e3
Removing intermediate container e5168f00e8e3
 ---> 334f8743d597
Step 15/29 : RUN go install -v std
 ---> Running in 53d805956be4
os/user
net
archive/tar
internal/testenv
plugin
vendor/golang_org/x/net/lex/httplex
crypto/x509
vendor/golang_org/x/net/proxy
net/textproto
log/syslog
crypto/tls
vendor/golang_org/x/net/internal/nettest
vendor/golang_org/x/net/nettest
net/smtp
net/http/httptrace
net/http
net/http/cgi
net/http/cookiejar
net/http/httptest
net/http/fcgi
net/http/httputil
net/rpc
net/rpc/jsonrpc
Removing intermediate container 53d805956be4
 ---> 184515245325
Step 16/29 : RUN go get github.com/Masterminds/glide
 ---> Running in d11abdc40749
Removing intermediate container d11abdc40749
 ---> aa1deb84f46b
Step 17/29 : ENV GLIDE_HOME /home/user/.glide
 ---> Running in 453a8f0b8485
Removing intermediate container 453a8f0b8485
 ---> 68a1ea79ac79
Step 18/29 : RUN go get github.com/golang/dep/cmd/dep
 ---> Running in 3b633d04337c
Removing intermediate container 3b633d04337c
 ---> 07ff8acf1ba7
Step 19/29 : RUN go get github.com/onsi/ginkgo/ginkgo
 ---> Running in e437ee3baafc
Removing intermediate container e437ee3baafc
 ---> cb331ff3b11a
Step 20/29 : RUN go get -u -d github.com/alecthomas/gometalinter &&     cd /go/src/github.com/alecthomas/gometalinter &&     git checkout cc4415ed09f7073d595ee504cad4d98b71a3038e &&     go install github.com/alecthomas/gometalinter
 ---> Running in 1c14ef4243d5
Note: checking out 'cc4415ed09f7073d595ee504cad4d98b71a3038e'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by performing another checkout.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -b with the checkout command again. Example:

  git checkout -b <new-branch-name>

HEAD is now at cc4415e... Add back support for predefined patterns in linter definitions.
Removing intermediate container 1c14ef4243d5
 ---> d312d900ee06
Step 21/29 : RUN ln -s `which gometalinter` /usr/local/bin/gometalinter
 ---> Running in cd26b70aa8a8
Removing intermediate container cd26b70aa8a8
 ---> 48ac0384a02b
Step 22/29 : RUN gometalinter --install
 ---> Running in 82b53d42b22b
Installing:
  deadcode
  dupl
  errcheck
  gas
  goconst
  gocyclo
  goimports
  golint
  gosimple
  gotype
  gotypex
  ineffassign
  interfacer
  lll
  maligned
  megacheck
  misspell
  safesql
  staticcheck
  structcheck
  unconvert
  unparam
  unused
  varcheck
  vet
Removing intermediate container 82b53d42b22b
 ---> f011e4985a0c
Step 23/29 : RUN go get github.com/pmezard/licenses
 ---> Running in bf1559e4458a
Removing intermediate container bf1559e4458a
 ---> 050cfe70363e
Step 24/29 : RUN go get github.com/wadey/gocovmerge
 ---> Running in 5de2f841c4a2
Removing intermediate container 5de2f841c4a2
 ---> 61cec8b9a5c7
Step 25/29 : RUN go get github.com/mikefarah/yaml
 ---> Running in 63c6069de155
Removing intermediate container 63c6069de155
 ---> 4abfe5d37413
Step 26/29 : RUN go get -u -d github.com/fasaxc/goveralls &&     cd /go/src/github.com/fasaxc/goveralls &&     git checkout tags/v0.0.1-smc &&     go install github.com/fasaxc/goveralls
 ---> Running in f269762f7d79
Note: checking out 'tags/v0.0.1-smc'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by performing another checkout.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -b with the checkout command again. Example:

  git checkout -b <new-branch-name>

HEAD is now at e305190... Only send a service job ID if it is provided.
Removing intermediate container f269762f7d79
 ---> 846343476c86
Step 27/29 : RUN chmod -R 777 $GOPATH
 ---> Running in cfe4b2d9a2b8
Removing intermediate container cfe4b2d9a2b8
 ---> 1157027cbe92
Step 28/29 : COPY entrypoint.sh /usr/local/bin/entrypoint.sh
 ---> 9793aa471179
Step 29/29 : ENTRYPOINT ["/sbin/tini", "--", "/usr/local/bin/entrypoint.sh"]
 ---> Running in fd9747af0596
Removing intermediate container fd9747af0596
 ---> 186032facb14
Successfully built 186032facb14
Successfully tagged calico/go-build:latest-ppc64le
</details>

complete log - https://gist.github.com/mkumatag/6c26ea00b0236279da6b82a23f857a25

mkumatag commented:
and configuration looks like below:

```shell
root@localhost:~/go-build# uname -a
Linux localhost 4.4.0-116-generic #140-Ubuntu SMP Mon Feb 12 21:23:04 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
root@localhost:~/go-build# docker version
Client:
 Version:	18.03.0-ce
 API version:	1.37
 Go version:	go1.9.4
 Git commit:	0520e24
 Built:	Wed Mar 21 23:10:01 2018
 OS/Arch:	linux/amd64
 Experimental:	false
 Orchestrator:	swarm

Server:
 Engine:
  Version:	18.03.0-ce
  API version:	1.37 (minimum version 1.12)
  Go version:	go1.9.4
  Git commit:	0520e24
  Built:	Wed Mar 21 23:08:31 2018
  OS/Arch:	linux/amd64
  Experimental:	false
root@localhost:~/go-build# cat /etc/os-release
NAME="Ubuntu"
VERSION="16.04.4 LTS (Xenial Xerus)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 16.04.4 LTS"
VERSION_ID="16.04"
HOME_URL="http://www.ubuntu.com/"
SUPPORT_URL="http://help.ubuntu.com/"
BUG_REPORT_URL="http://bugs.launchpad.net/ubuntu/"
VERSION_CODENAME=xenial
UBUNTU_CODENAME=xenial
root@localhost:~/go-build#
```

mkumatag commented:
I created another virtual machine with same kernel version able to build all the images, though cpu is different,

```
processor	: 0
vendor_id	: GenuineIntel
cpu family	: 6
model		: 45
model name	: Intel(R) Xeon(R) CPU E5-2690 0 @ 2.90GHz
stepping	: 2
microcode	: 0x710
cpu MHz		: 2900.000
cache size	: 20480 KB
physical id	: 0
siblings	: 1
core id		: 0
cpu cores	: 1
apicid		: 0
initial apicid	: 0
fpu		: yes
fpu_exception	: yes
cpuid level	: 13
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts mmx fxsr sse sse2 ss syscall nx rdtscp lm constant_tsc arch_perfmon pebs bts nopl xtopology tsc_reliable nonstop_tsc cpuid aperfmperf pni pclmulqdq ssse3 cx16 pcid sse4_1 sse4_2 x2apic popcnt tsc_deadline_timer aes xsave avx hypervisor lahf_lm epb pti retpoline tsc_adjust dtherm ida arat pln pts
bugs		: cpu_meltdown spectre_v1 spectre_v2
bogomips	: 5800.00
clflush size	: 64
cache_alignment	: 64
address sizes	: 42 bits physical, 48 bits virtual
power management:

processor	: 1
vendor_id	: GenuineIntel
cpu family	: 6
model		: 45
model name	: Intel(R) Xeon(R) CPU E5-2690 0 @ 2.90GHz
stepping	: 2
microcode	: 0x710
cpu MHz		: 2900.000
cache size	: 20480 KB
physical id	: 2
siblings	: 1
core id		: 0
cpu cores	: 1
apicid		: 2
initial apicid	: 2
fpu		: yes
fpu_exception	: yes
cpuid level	: 13
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts mmx fxsr sse sse2 ss syscall nx rdtscp lm constant_tsc arch_perfmon pebs bts nopl xtopology tsc_reliable nonstop_tsc cpuid aperfmperf pni pclmulqdq ssse3 cx16 pcid sse4_1 sse4_2 x2apic popcnt tsc_deadline_timer aes xsave avx hypervisor lahf_lm epb pti retpoline tsc_adjust dtherm ida arat pln pts
bugs		: cpu_meltdown spectre_v1 spectre_v2
bogomips	: 5800.00
clflush size	: 64
cache_alignment	: 64
address sizes	: 42 bits physical, 48 bits virtual
power management:
```

For testing purpose, I even created sample [Travis job](https://travis-ci.org/mkumatag/go-build/builds/351149366?utm_source=github_status&utm_medium=notification) and ran it, and even that passed. I strongly suspect something is wrong with your setup!

djlwilder commented:
I am seeing the same issue as @tomdee.
@mkumatag I can give you access to my environment (will send private email),  maybe we can figure this out.

But what really matters is, will it build on semaphore's environment?  should we try that?  Ci currently runs "make test" you can hack the Makefile in a separate test PR so that make test builds all archs then we can see if it works.

fasaxc commented:
FWIW, the build seems flaky on my machine.  One failure, one success, one hang :-/  Having tried a go cross build like this before for a personal project, I found that using `GOMAXPROCS=1 go build -p 1 ...` seemed to work better with qemu-static

fasaxc commented:
I'll add a semaphore job that runs the new targets if they're there...

djlwilder commented:
The semaphore run hung in the arm64 build.  
Step 17/28 : RUN go get github.com/golang/dep/cmd/dep
 ---> Running in fc2fa688924f
execution expired

deitch commented:
And the most recent one has it hanging on a different `go get`:

```
Step 15/28 : RUN go get github.com/Masterminds/glide
 ---> Running in 5cd65f453871
execution expired
```

I suspect that it is a `git` issue, since `go get` like that just forks `git` and waits for stdout/stderr.

Any way we can add debugging to see what it is doing, or possibly get in while it hangs and see processes?

djlwilder commented:
I built a simple container with just qemu-user-ppc64le and go.  Even a simple "go build hello.go" fails in this environment.   I tested @fasaxc suggestion of using GOMAXPROCS=1 go build -p 1,  that avoided the problem, no failures.  Thanks for the  suggestion @fasaxc!  I went back and add the that change to every go command in Dockerfile.ppc64le  and Dockerfile.arm64 with that change both built ok.  Qemu looks to be racy with signals and lacks correct thread implementation, but we have a workaround.

djlwilder commented:
This qemu bug seams to describe the issue we are seeing.  Although it claims to be fixed upstream.
https://bugs.launchpad.net/qemu/+bug/1098729

tomdee commented:
As an alternative, I've tried adding a docker build arg to tell it to just use one CPU and that seems to do the trick too. It's running on semaphore now :crossed_fingers: https://semaphoreci.com/calico/go-build/branches/mkumatag-cross_build/builds/1

mkumatag commented:
Thanks, @djlwilder @tomdee for debugging further. I see above semaphoreci build is passed with https://github.com/projectcalico/go-build/commit/1b15ad08d4393f506491c42bdd9945af29027574. I'll add this PR with restriction of docker build with single cpu.

deitch commented:
Nice @djlwilder . I would have gone down the git rabbit hole and come out confused. 

> I went back and add the that change to every go command in Dockerfile.ppc64le and Dockerfile.arm64 with that change both built ok

Or just adding an `ENV` line to the `Dockerfile` would do it. Of course, it means slowing down builds for non-cross-compile, but it is so fast anyways...

mkumatag commented:
Hurrah! build is passing now.!

There is a scope for improving the performance - as we are already using the multi-stage docker build, we can build all those go utilities in STAGE 0 in the host architecture and then copy to the second STAGE directly. I'm deferring this work to another PR.

djlwilder commented:
go-build images are not built very often.  I don't think speeding up the build is a priority.  Thanks to everyone for working on this.  I am looking forward to running "docker pull calico/go-build" in my ppc64le environment!!.

mkumatag commented:
@tomdee @fasaxc I have updated the PR with following things:
- Rebased with the latest code
- Using the manifest-tool from container @deitch 

PTAL and review it.

fasaxc commented on Makefile:
```diff
@@ -19,39 +19,50 @@ VERSION ?= latest
+calico/go-build: register
 	# Make sure we re-pull the base image to pick up security fixes.
-	docker build --pull -t $(ARCHIMAGE) -f $(DOCKERFILE) .
+	# Limit the build to use only one CPU, This helps to work around qemu bugs such as https://bugs.launchpad.net/qemu/+bug/1098729
```
Can you limit the `--cpu-set` arg to only apply for cross-builds?  I'd rather not slow down the amd64 build if possible.

mkumatag replied:
I'll address this in another PR. -Thanks.

fasaxc commented:
Let's get this merged so it doesn't have to be rebased again!  Happy for you to follow up with those tweaks, @mkumatag
